drivers/arm/serial_usart.c

```diff
@@ -0,0 +1,198 @@
+#include "quantum.h"
+#include "serial.h"
+#include "printf.h"
+
+#include "ch.h"
+#include "hal.h"
+
+#ifndef USE_GPIOV1
+// The default PAL alternate modes are used to signal that the pins are used for USART
+#    ifndef USART_TX_PAL_MODE
+#        define USART_TX_PAL_MODE 7
+#    endif
+#endif
+
+#ifndef USART_DRIVER
+#    define USART_DRIVER SD1
+#endif
+
+#ifndef SELECT_SOFT_SERIAL_SPEED
+#    define SELECT_SOFT_SERIAL_SPEED 1
+// TODO: correct speeds...
+//  0: about 189kbps (Experimental only)
+//  1: about 137kbps (default)
+//  2: about 75kbps
+//  3: about 39kbps
+//  4: about 26kbps
+//  5: about 20kbps
+#endif
+
+// Serial pulse period in microseconds. At the moment, going lower than 12 causes communication failure
+#if SELECT_SOFT_SERIAL_SPEED == 0
+#    define USART_SPEED 230400
+#elif SELECT_SOFT_SERIAL_SPEED == 1
+#    define USART_SPEED 115200
+#elif SELECT_SOFT_SERIAL_SPEED == 2
+#    define USART_SPEED 57600
+#elif SELECT_SOFT_SERIAL_SPEED == 3
+#    define USART_SPEED 38400
+#elif SELECT_SOFT_SERIAL_SPEED == 4
+#    define USART_SPEED 19200
+#elif SELECT_SOFT_SERIAL_SPEED == 5
+#    define USART_SPEED 14400
+#else
+#    error invalid SELECT_SOFT_SERIAL_SPEED value
+#endif
+
+#define TIMEOUT 100
+
+static inline void sdClear(SerialDriver* driver) {
+    while (sdGetTimeout(driver, TIME_IMMEDIATE) != MSG_TIMEOUT) {
+        // Do nothing with the data
+    }
+}
+
+static inline msg_t sdPutHalfDuplex(SerialDriver* driver, uint8_t data) {
+    msg_t ret = sdPut(driver, data);
+    sdGet(driver);
```

Comment:
Lets add `// Half duplex requires us to read back the data we just wrote - just throw it away` to match the other functions.